Dockerfile.cross

```diff
@@ -0,0 +1,297 @@
+# syntax=docker/dockerfile:1
+
+ARG BASE_VARIANT=bullseye
+ARG GO_VERSION=1.18.1
+ARG XX_VERSION=1.1.0
+
+ARG CGO_ENABLED
+ARG GO_LINKMODE=static
+
+ARG CONTAINERD_VERSION=v1.6.2
+ARG RUNC_VERSION=v1.1.1
+ARG VPNKIT_VERSION=0.5.0
+ARG TINI_VERSION=v0.19.0
+ARG ROOTLESSKIT_VERSION=1920341cd41e047834a21007424162a2dc946315
+ARG CONTAINERUTILITY_VERSION=aa1ba87e99b68e0113bd27ec26c60b88f9d4ccd9
+ARG GOWINRES_VERSION=v0.2.3
+
+# cross compilation helper
+FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
+
+# dummy stage to make sure the image is built for unsupported deps
+FROM --platform=$BUILDPLATFORM busybox AS build-dummy
+RUN mkdir -p /out
+FROM scratch AS binary-dummy
+COPY --from=build-dummy / /
+
+# FIXME: cannot build with alpine variant atm
+FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-${BASE_VARIANT} AS build-base-alpine
+ENV GO111MODULE=off
+COPY --from=xx / /
+RUN apk add --no-cache bash cmake file git make
+WORKDIR /go/src/github.com/docker/docker
+
+FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-${BASE_VARIANT} AS build-base-bullseye
+ENV GO111MODULE=off
+ENV DEBIAN_FRONTEND=noninteractive
+COPY --from=xx / /
+# bullseye-backports for cmake
+RUN echo "deb https://deb.debian.org/debian bullseye-backports main contrib non-free" >> /etc/apt/sources.list
+RUN apt-get update -y && apt-get install --no-install-recommends -y bash file git make
+RUN apt-get -y -t bullseye-backports install cmake
+WORKDIR /go/src/github.com/docker/docker
+
+# go-winres
+FROM build-base-${BASE_VARIANT} AS gowinres
+ARG GOWINRES_VERSION
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
+    GOBIN=/out GO111MODULE=on go install "github.com/tc-hib/go-winres@${GOWINRES_VERSION}" \
+    && /out/go-winres --help
+
+# containerd
+FROM build-base-${BASE_VARIANT} AS containerd-src
+WORKDIR /usr/src
+RUN git clone https://github.com/containerd/containerd.git containerd
+
+FROM build-base-${BASE_VARIANT} AS containerd-base-static
+WORKDIR /go/src/github.com/containerd/containerd
+ENV CGO_ENABLED=1
+ENV BUILDTAGS="netgo osusergo static_build"
+ENV EXTRA_LDFLAGS='-linkmode external -extldflags "-fno-PIC -static"'
+ENV GO111MODULE=off
+
+FROM build-base-${BASE_VARIANT} AS containerd-base-dynamic
+WORKDIR /go/src/github.com/containerd/containerd
+ENV GO111MODULE=off
+
+FROM containerd-base-${GO_LINKMODE} AS containerd-base-alpine
+ARG TARGETPLATFORM
+RUN xx-apk add --no-cache musl-dev gcc g++ pkgconf btrfs-progs-dev libsecret-dev && xx-go --wrap
```

Comment:
we could add a mount cache here for alpine pkgs s well as for the debian variant